sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -622,31 +624,47 @@ case class Range(
 /**
  * This is a Group by operator with the aggregate functions and projections.
  *
- * @param groupingExpressions expressions for grouping keys
- * @param aggregateExpressions expressions for a project list, which could contain
- *                             [[AggregateExpression]]s.
+ * @param groupingExpressions Expressions for grouping keys.
+ * @param aggrExprWithGroupingRefs Expressions for a project list, which could contain
+ *                                 [[AggregateExpression]]s and [[GroupingExprRef]]s.
+ * @param child The child of the aggregate node.
+ * @param enforceGroupingReferences If [[aggrExprWithGroupingRefs]] should contain
+ *                                  [[GroupingExprRef]]s.
+ *
+ * [[aggrExprWithGroupingRefs]] without aggregate functions can contain [[GroupingExprRef]]
+ * expressions to refer to complex grouping expressions in [[groupingExpressions]]. These references
+ * ensure that optimization rules don't change the aggregate expressions to invalid ones that no
+ * longer refer to any grouping expressions and also simplify the expression transformations on the
+ * node (need to transform the expression only once).
  *
- * Note: Currently, aggregateExpressions is the project list of this Group by operator. Before
- * separating projection from grouping and aggregate, we should avoid expression-level optimization
- * on aggregateExpressions, which could reference an expression in groupingExpressions.
- * For example, see the rule [[org.apache.spark.sql.catalyst.optimizer.SimplifyExtractValueOps]]
+ * For example, in the following query Spark shouldn't optimize the aggregate expression
+ * `Not(IsNull(c))` to `IsNotNull(c)` as the grouping expression is `IsNull(c)`:
+ * SELECT not(c IS NULL)
+ * FROM t
+ * GROUP BY c IS NULL
+ * Instead, the aggregate expression should contain `Not(GroupingExprRef(0))`.
  */
-case class Aggregate(
+case class AggregateWithGroupingReferences(
```

Comment:
The reason for this rename is to keep the semantics of `Aggregate([grouping expressions], [aggregate expressions without grouping references], [child])` extractor, but redefining the extractor of a case class doesn't seem to work.

Reply:
This change is much bigger than I expect. What's the reason here? I was expecting something like
1. A rule that runs at the beginning of optimizer, to create the grouping refs.
2. Restore grouping refs to the original expressions when converting logical aggregate to phyiscal versions.

Reply:
During the optimization phase some of the rules need the "expanded" version of aggregate expressions (references are restored to original expressions). This is why I wanted to keep the meaning of `Aggregate.aggregateExpressions` and the second argument of the `Aggregate.unapply()` extractor. Also when an optimizer rule creates an aggregate with the `Aggregete.apply()` it should automatically insert references the grouping expressions.
I think it would be error prone if the optimizer rules need to keep in mind that they need to convert/restore these references.

Reply:
> some of the rules need the "expanded" version of aggregate expressions (references are restored to original expressions)

do you have some examples?

Reply:
I think you are right, only a very few of them seems to really need expanded aggregate expressions ( `RemoveRedundantAggregates`, `RewriteCorrelatedScalarSubquery`). In https://github.com/apache/spark/commit/c2ba80457bd86d11ad26311bbc3c42607f33b19a I dropped handling group references transparently.